packages/bitcore-node/src/services/pruning.ts

```diff
@@ -112,144 +121,173 @@ export class PruningService {
-                }
-                if (coin.spentTxid) {
-                  spentTxids.add(coin.spentTxid);
-                  if (spentTxids.size > DESCENDANT_LIMIT) {
-                    logger.warn(`${tx.txid} has too many decendants`);
-                    return cb();
-                  }
-                }
-              }
-              spentTxids.add(tx.txid);
-              const uniqueTxids = Array.from(spentTxids);
-              await this.clearInvalid(uniqueTxids);
-              logger.info(`Invalidated tx ${tx.txid} and ${spentTxids.size - 1} dependent txs`);
-              cb();
+    try {
+      if (this.runningInvalid) {
+        return;
+      }
+      this.runningInvalid = true;
+
+      if (Date.now() - this.lastRunTimeInvalid < INV_INTERVAL_MINS * ONE_MIN) {
+        return;
+      }
+
+      const count = await this.coinModel.collection.countDocuments({ chain, network, mintHeight: SpentHeightIndicators.pending });
+      logger.info(`Found ${count} pending ${chain}:${network} TXOs`);
+
+      // Note, realCount <= count since the coinStream returns coins dynamically and related coins are updated.
+      // The caveat is that new coins could be added to the stream while we are iterating over it.
+      let realCount = 0;
+      let invalidCount = 0;
+      const seen = new Set<string>();
+      const voutStream = this.coinModel.collection.find({ chain, network, mintHeight: SpentHeightIndicators.pending }).batchSize(1);
+      for await (const vout of voutStream) {
+        if (this.stopping) {
+          break;
+        }
+        realCount++;
+        if (seen.has(vout.mintTxid)) {
+          continue;
+        }
+        seen.add(vout.mintTxid);
+        const tx = await this.transactionModel.collection.findOne({ chain, network, txid: vout.mintTxid });
+        if (!tx) {
+          logger.error(`Coin ${vout.mintTxid} has no corresponding tx`);
+          continue;
+        }
+        if (tx.replacedByTxid) {
+          if (await this.invalidateTx(chain, network, tx)) {
+            invalidCount++;
+          }
+        } else {
+          const vins = await this.coinModel.collection.find({ chain, network, spentTxid: vout.mintTxid }).toArray();
+          const vinTxs = await this.transactionModel.collection.find({ chain, network, txid: { $in: vins.map(vin => vin.mintTxid) } }).toArray();
+          for (const tx of vinTxs) {
+            if (tx.replacedByTxid) {
+              if (await this.invalidateTx(chain, network, tx)) {
+                invalidCount++;
+              };
             }
-          })
-        )
-        .on('finish', resolve)
-        .on('error', reject);
-    });
+          }
+        }
+      }
+      logger.info(`Invalidated ${invalidCount} (processed ${realCount}) pending TXOs for ${chain}:${network}`);
+      this.lastRunTimeInvalid = Date.now();
+    } catch (err: any) {
+      logger.error(`Error processing invalid txs: ${err.stack || err.message || err}`);
```

Comment:
This is called from a try block. If you don't throw an error, then you'll never receive the catch behavior specified where it's called.

Reply:
I just want to log and continue